### PR TITLE
Enable nonzero k_point in adjoint solver

### PR DESCRIPTION
**Pull request by mochen4**
Fixes https://github.com/NanoComp/meep/issues/1667. The [test_adjoint_solver.py](https://github.com/NanoComp/meep/blob/master/python/tests/test_adjoint_solver.py) is updated correspondingly to include when ``k_point`` is ``False``, zero, and nonzero.

### CODE REVIEW

**Comment by mochen4** [comment]
The original setup of the ``EigenmodeCoefficient`` test in the [test_adjoint_solver.py](https://github.com/NanoComp/meep/blob/master/python/tests/test_adjoint_solver.py) has the source and monitor coincide with the sides of the PMLs, and extends into the PMLs of the other direction.  The adjoint gradients and directional derivates won't match if ``k_point`` is nonzero; if we use the original ``calculate_fd_gradient()`` in [optimization_problem.py](https://github.com/NanoComp/meep/blob/master/python/adjoint/optimization_problem.py) that calculates the fd_gradients in unit directions, the fields become ``nan``. (The forward and adjoint run are both okay, but the fields are ``nan`` after single pixel is perturbed during ``calculate_fd_gradient()``.)

Using the setup of the [tutorial example](https://github.com/NanoComp/meep/blob/master/python/examples/adjoint_optimization/01-Introduction.ipynb), the gradients agree with nonzero ``k_point``.

If the source and monitor don't extend into the PMLs, then the adjoint gradients will match the directional derivates, but the fields are still ``nan`` when using ``calculate_fd_gradient()``. If we then moved the source and monitor a little bit away from the PMLs, everything will work fine; so I further modified the test this way.

**Comment by mochen4** [comment]
In the context of the PR https://github.com/NanoComp/meep/pull/1704, the factor of 2 is needed whenever ``force_complex_fields=False``. However, when ``k_point`` is nonzero,  we may not need the factor even if we don't force complex fields.

I wonder if there are other cases where we don't need the factor even if ``force_complex_fields=False``; and what will be the best way to check whether indeed only real fields are used.

**Comment by oskooi** [comment]
Looks like the failing test is `test_adjoint_solver.py` for the case in which Meep is compiled with single precision (#1698). Perhaps you need to just lower the tolerance some place when `mp.is_single_precision()` is true?

**Comment by mochen4** [comment]
Thanks Ardavan! I lowered the tolerance for single precision, and it works!

File: python/adjoint/objective.py [hunk 1]
```diff
@@ -92,6 +92,8 @@ def _adj_src_scale(self, include_resolution=True):
         else:
             # multi frequency simulations
             scale = dV * iomega / adj_src_phase
+        if ((not self.sim.k_point) or self.sim.k_point==mp.Vector3()) and not self.sim.force_complex_fields:
```
Comment (stevengj) [inline python/adjoint/objective.py:95]: We should really check `fields.is_real`, assuming the `fields` object exists here.

Comment (stevengj) [reply python/adjoint/objective.py:95]: Even simpler, check `self.use_real()`

Comment (stevengj) [reply python/adjoint/objective.py:95]: `use_real` is currently an internal function of `init_sim`, which can just be split into a separate method, and probably renamed to `using_real_fields()`

**Pull request merged**